### ISSUE TITLE
Macros: multi-stage output events

### ISSUE THREAD

**Issue by Muirium** [pqrs-org/Karabiner-Elements#2711]
This is probably a feature request. I have searched Karabiner's documentation and other people's rules for macros but can't find anything about it.

What I want to achieve with macros is stuff like this, on a single trigger keystroke:

- Type the À character: press Option + ` _then_ press Shift + A
- Select the current word: press Option + Left arrow _then_ press Shift + Option + Right arrow
- Type ABcd%^: Hold Shift _then_ press A _then_ B _then_ release Shift _then_ press C _then_ D _then_ hold Shift _then_ press 5 _then_ 6
- Combining the above to do magic like (( wrap )) the current word in [a]tags[/b] instantly with a single keystroke!
- And throwing the clipboard contents into the mix, for things like [url=pasteboard]selection[/url]. Boom.

In a nutshell: "from": events triggering whole chains of "to": events.

I do lots of macro stuff like this with [Soarer's Converter](https://deskthority.net/viewtopic.php?f=7&t=2510&start=) and it can be tremendously useful. I'd love to be able to do it across all my keyboards, including my MacBook Air's own. Are sequences simply beyond Karabiner's scope?

(Karabiner really is a lot of fun, once you [get into writing your own](https://deskthority.net/viewtopic.php?f=7&t=25461) complex rules. Brilliant stuff! My biggest thanks to Tekezo and everyone who works on this project. Love it to bits!)

**Comment by Muirium** [comment]
I'm pleased to report Karabiner can totally do this!

Finally occurred to me to just try putting multiple `key_code` statements into the `to` definitions. There's a single example of it somewhere in Karabiner's documentation, just shown in passing, typing out the word `hello`. Once I saw that I was at the races!

So I've made a bunch. Typing out simple strings is just the start of it. What you'll find here is word selection and paragraph selection macros—which use sequences of modifier keys to do their work—and wrapping macros which throw the clipboard into the mix with Command + X and V. 

Wrapping (braces) etc. around selected text using the clipboard to store the text isn't the cleanest way, I know, but it works; and does so **everywhere!** These macros all work in every app across the Mac thanks to Karabiner; much more reliably than [Cocoa Keybindings](https://www.hcs.harvard.edu/~jrus/site/selectors.html) which are elegant but often not supported at all in an app. I use [Jumpcut](https://snark.github.io/jumpcut/) for clipboard history, so destructive cut and paste is fine but not great. Is there a cleaner way, I wonder?
```
{
	"title": "🦾 General Macros",
	"rules": [
		{
			"description": "🤖 Select Current Word: ⌃ Control + ⌥ Option + ← Left or → Right ⇨ Select Word",
			"manipulators": [
				{
					"from": {
						"key_code": "left_arrow",
						"modifiers": {
							"mandatory": [
								"option",
								"control"
							]
						}
					},
					"to": [
						{
							"key_code": "left_arrow",
							"modifiers": ["left_option"]
						},
						{
						    "key_code": "right_arrow",
						    "modifiers": ["left_option", "left_shift"]
						}
					],
					"type": "basic"
				},
				{
					"from": {
						"key_code": "right_arrow",
						"modifiers": {
							"mandatory": [
								"option",
								"control"
							]
						}
					},
					"to": [
						{
							"key_code": "right_arrow",
							"modifiers": ["left_option"]
						},
						{
						    "key_code": "left_arrow",
						    "modifiers": ["left_option", "left_shift"]
						}
					],
					"type": "basic"
				}
			]
		},
		{
			"description": "🤖 Select Current Para: ⌃ Control + ⌥ Option + ↑ Up or ↓ Down ⇨ Select Paragraph",
			"manipulators": [
				{
					"from": {
						"key_code": "up_arrow",
						"modifiers": {
							"mandatory": [
								"option",
								"control"
							]
						}
					},
					"to": [
					    {
							"key_code": "up_arrow"
						},
						{
							"key_code": "up_arrow",
							"modifiers": ["left_option"]
						},
						{
						    "key_code": "down_arrow",
						    "modifiers": ["left_option", "left_shift"]
						}
					],
					"type": "basic"
				},
				{
					"from": {
						"key_code": "down_arrow",
						"modifiers": {
							"mandatory": [
								"option",
								"control"
							]
						}
					},
					"to": [
						{
							"key_code": "down_arrow"
						},
						{
							"key_code": "down_arrow",
							"modifiers": ["left_option"]
						},
						{
						    "key_code": "up_arrow",
						    "modifiers": ["left_option", "left_shift"]
						}
					],
					"type": "basic"
				}
			]
		},
		{
			"description": "🤖 Wrap Selection: ⌃ Control + ⌥ Option + '\"(*!? ⇨ Paired Bracket Wraps",
			"manipulators": [
				{
					"from": {
						"key_code": "open_bracket",
						"modifiers": {
							"mandatory": [
								"option",
								"control"
							]
						}
					},
					"to": [
						{
						    "key_code": "x",
						    "modifiers": ["left_command"]
						},
						{
						    "key_code": "open_bracket"
						},
						{
						    "key_code": "v",
						    "modifiers": ["left_command"]
						},
						{
						    "key_code": "close_bracket"
						}
					],
					"type": "basic"
				},
				{
					"from": {
						"key_code": "open_bracket",
						"modifiers": {
							"mandatory": [
								"option",
								"control",
								"shift"
							]
						}
					},
					"to": [
						{
						    "key_code": "x",
						    "modifiers": ["left_command"]
						},
						{
						    "key_code": "open_bracket",
						     "modifiers": ["left_shift"]
						},
						{
						    "key_code": "v",
						    "modifiers": ["left_command"]
						},
						{
						    "key_code": "close_bracket",
						     "modifiers": ["left_shift"]
						}
					],
					"type": "basic"
				},
				{
					"from": {
						"key_code": "quote",
						"modifiers": {
							"mandatory": [
								"option",
								"control"
							]
						}
					},
					"to": [
						{
						    "key_code": "x",
						    "modifiers": ["left_command"]
						},
						{
						    "key_code": "quote"
						},
						{
						    "key_code": "v",
						    "modifiers": ["left_command"]
						},
						{
						    "key_code": "quote"
						}
					],
					"type": "basic"
				},
				{
					"from": {
						"key_code": "quote",
						"modifiers": {
							"mandatory": [
								"option",
								"control",
								"shift"
							]
						}
					},
					"to": [
						{
						    "key_code": "x",
						    "modifiers": ["left_command"]
						},
						{
						    "key_code": "quote",
						     "modifiers": ["left_shift"]
						},
						{
						    "key_code": "v",
						    "modifiers": ["left_command"]
						},
						{
						    "key_code": "quote",
						     "modifiers": ["left_shift"]
						}
					],
					"type": "basic"
				},
				{
					"from": {
						"key_code": "9",
						"modifiers": {
							"mandatory": [
								"option",
								"control",
								"shift"
							]
						}
					},
					"to": [
						{
						    "key_code": "x",
						    "modifiers": ["left_command"]
						},
						{
						    "key_code": "9",
						     "modifiers": ["left_shift"]
						},
						{
						    "key_code": "v",
						    "modifiers": ["left_command"]
						},
						{
						    "key_code": "0",
						     "modifiers": ["left_shift"]
						}
					],
					"type": "basic"
				},
				{
					"from": {
						"key_code": "1",
						"modifiers": {
							"mandatory": [
								"option",
								"control",
								"shift"
							]
						}
					},
					"to": [
						{
						    "key_code": "x",
						    "modifiers": ["left_command"]
						},
						{
						    "key_code": "1",
						     "modifiers": ["left_option"]
						},
						{
						    "key_code": "v",
						    "modifiers": ["left_command"]
						},
						{
						    "key_code": "1",
						     "modifiers": ["left_shift"]
						}
					],
					"type": "basic"
				},
				{
					"from": {
						"key_code": "slash",
						"modifiers": {
							"mandatory": [
								"option",
								"control",
								"shift"
							]
						}
					},
					"to": [
						{
						    "key_code": "x",
						    "modifiers": ["left_command"]
						},
						{
						    "key_code": "slash",
						     "modifiers": ["left_shift", "left_option"]
						},
						{
						    "key_code": "v",
						    "modifiers": ["left_command"]
						},
						{
						    "key_code": "slash",
						     "modifiers": ["left_shift"]
						}
					],
					"type": "basic"
				},
				{
					"from": {
						"key_code": "8",
						"modifiers": {
							"mandatory": [
								"option",
								"control",
								"shift"
							]
						}
					},
					"to": [
						{
						    "key_code": "x",
						    "modifiers": ["left_command"]
						},
						{
						    "key_code": "slash"
						},
						{
						    "key_code": "8",
						    "modifiers": ["left_shift"]
						},
						{
						    "key_code": "spacebar"
						},
						{
						    "key_code": "v",
						    "modifiers": ["left_command"]
						},
						{
						    "key_code": "spacebar"
						},
						{
						    "key_code": "8",
						    "modifiers": ["left_shift"]
						},
						{
						    "key_code": "slash"
						}
					],
					"type": "basic"
				},
				{
					"from": {
						"key_code": "g",
						"modifiers": {
							"mandatory": [
								"option",
								"control"
							]
						}
					},
					"to": [
						{
						    "key_code": "x",
						    "modifiers": ["left_command"]
						},
						{
						    "key_code": "open_bracket"
						},
						{
						    "key_code": "i"
						},
						{
						    "key_code": "m"
						},
						{
						    "key_code": "g"
						},
						{
						    "key_code": "close_bracket"
						},
						{
						    "key_code": "v",
						    "modifiers": ["left_command"]
						},
						{
						    "key_code": "open_bracket"
						},
						{
						    "key_code": "slash"
						},
						{
						    "key_code": "i"
						},
						{
						    "key_code": "m"
						},
						{
						    "key_code": "g"
						},
						{
						    "key_code": "close_bracket"
						}
					],
					"type": "basic"
				},
				{
					"from": {
						"key_code": "k",
						"modifiers": {
							"mandatory": [
								"option",
								"control"
							]
						}
					},
					"to": [
						{
						    "key_code": "x",
						    "modifiers": ["left_command"]
						},
						{
						    "key_code": "open_bracket"
						},
						{
						    "key_code": "u"
						},
						{
						    "key_code": "r"
						},
						{
						    "key_code": "l"
						},
						{
						    "key_code": "equal_sign"
						},
						{
						    "key_code": "close_bracket"
						},
						{
						    "key_code": "v",
						    "modifiers": ["left_command"]
						},
						{
						    "key_code": "open_bracket"
						},
						{
						    "key_code": "slash"
						},
						{
						    "key_code": "u"
						},
						{
						    "key_code": "r"
						},
						{
						    "key_code": "l"
						},
						{
						    "key_code": "close_bracket"
						}
					],
					"type": "basic"
				}
			]
		}
    ]
}
```

**Comment by Muirium** [comment]
Here's an example of speeding up dead keys for diacritical marks. I want to type àèìòù as easily as holding down Shift to type AEIOU. Nò pròblèmò!

```
{
	"title": "🏴󠁧󠁢󠁳󠁣󠁴󠁿 Gaelic Macros",
	"rules": [
		{
			"description": "🏴󠁧󠁢󠁳󠁣󠁴󠁿 Gaelic Stràc Macros: 🦾 ⌃ Control + ⌥ Option + aeiou ⇨ àèìòù",
			"manipulators": [
				{
					"from": {
						"key_code": "a",
						"modifiers": {
							"mandatory": [
								"option",
								"control"
							]
						}
					},
					"to": [
						{
							"key_code": "grave_accent_and_tilde",
							"modifiers": ["left_option"]
						},
						{
						    "key_code": "a"
						}
					],
					"type": "basic"
				},
				{
					"from": {
						"key_code": "a",
						"modifiers": {
							"mandatory": [
								"option",
								"control",
								"shift"
							]
						}
					},
					"to": [
						{
							"key_code": "grave_accent_and_tilde",
							"modifiers": ["left_option"]
						},
						{
						    "key_code": "a",
						    "modifiers": ["left_shift"]
						}
					],
					"type": "basic"
				}
```
Again, a separate macro is necessary for à and À, because Shift gets in the way of the dead key sequence Karabiner's robotically pressing really quickly for me. Previously, every stràc that came up in Gaelic, I'd have to twiddle around with Option + ` then press the key. That made À a four key sequence. Much better when the robot's got your back!

I guess this issue is now closed. I'll mark it closed in a bit. 

I can imagine improvements to support macros: especially storing the pasteboard. But now I can see how Karabiner approaches this problem, I recognise I'm imagining un-Karabiner things!

**Comment by MuhammedZakir** [comment]
About clipboard problem:

You can save the current conent in main clipboard in a different clipboard, and move them back after seding `command+v`. See https://github.com/pqrs-org/KE-complex_modifications/issues/697#issuecomment-678677912 and `pbcopy/pbpaste` manpage.

**Comment by Muirium** [comment]
Oh, that's helpful! Thanks. Here it is in action:

```
{
	"title": "Test Rules",
	"rules": [
		{
			"description": "🤖 Wrap Selection: ⌃ Control + ⌥ Option + '\"(*!? ⇨ Paired Bracket Wraps",
			"manipulators": [
				{
					"from": {
						"key_code": "open_bracket",
						"modifiers": {
							"mandatory": [
								"option",
								"control",
								"command"
							]
						}
					},
					"to": [
					    {
						    "key_code": "x",
						    "modifiers": ["left_command"]
						},
						{
						    "shell_command": "pbpaste | pbcopy -pboard ruler"
						},
						{
						    "key_code": "open_bracket"
						},
						{
						    "shell_command": "pbpaste -pboard ruler"
						},
						{
						    "key_code": "v",
						    "modifiers": ["left_command"]
						},
						{
						    "key_code": "close_bracket"
						}
					],
					"type": "basic"
				}
			]
		}
	]
}
```
My only complaint is it still messes up clipboard history.

Ideally, I'd like to run these macros without messing up my clipboard or its history, given I can see that as a menu in [Jumpcut](https://snark.github.io/jumpcut/). I use my clipboard history a lot and would love to keep it clean while using these macros.

Any way to bypass the general clipboard entirely, and send / restore the selection straight to / from the ruler pasteboard instead? 

I keep dreaming of Karabiner manipulating the clipboard itself, rather than just the keyboard!

**Comment by MuhammedZakir** [comment]
Have you tried `control+k` (cut) and `contorl+y` shortucts? They use a different clipboard.

> Any way to bypass the general clipboard entirely, and send / restore the selection straight to / from the ruler pasteboard instead?

If you want to use `pbcopy` and `pbpaste`, then you can use AppleScript to insert text without messing the main clibpoard. Get clipboard content (simplest way is to use shell command) and then insert them using `keystroke`. I haven't tested it.

---

Your this `to` event
```
						    "shell_command": "pbpaste -pboard ruler"
```

sends the output to Karabiner which will discard them. I think your `to` events are wrong. It should be
```
					"to": [
						{
						    "shell_command": "pbpaste | pbcopy -pboard ruler"
						},
	  				        {
						    "key_code": "x",
						    "modifiers": ["left_command"]
						},
						{
						    "key_code": "open_bracket"
						},
						{
						    "key_code": "v",
						    "modifiers": ["left_command"]
						},
						{
						    "shell_command": "pbpaste -pboard ruler | pbcopy"
						},
						{
						    "key_code": "close_bracket"
						}
					],
```

**Comment by Muirium** [comment]
Control + K seems to use the main system clipboard here on M1 Big Sur. I tested with this:

```
{
	"title": "Test Rules",
	"rules": [
		{
			"description": "🤖 Wrap Selection: ⌃ Control + ⌥ Option + '\"(*!? ⇨ Paired Bracket Wraps",
			"manipulators": [
				{
					"from": {
						"key_code": "open_bracket",
						"modifiers": {
							"mandatory": [
								"option",
								"control",
								"command"
							]
						}
					},
					"to": [
					    {
						    "key_code": "k",
						    "modifiers": ["left_control"]
						},
						{
						    "key_code": "open_bracket"
						},
						{
						    "key_code": "y",
						    "modifiers": ["left_control"]
						},
						{
						    "key_code": "close_bracket"
						}
					],
					"type": "basic"
				}
			]
		}
	]
}
```
Running that macro on a text selection successfully wraps it with [brackets] but the text winds up in my clipboard history just as before. Strange.

**Comment by Muirium** [comment]
Okay, I think I get what you mean now! It's about saving the clipboard before I meddle with it. Here's my test of your suggested code:

```
{
	"title": "Test Rules",
	"rules": [
		{
			"description": "Test",
			"manipulators": [
				{
					"from": {
						"key_code": "open_bracket",
						"modifiers": {
							"mandatory": [
								"option",
								"control",
								"command"
							]
						}
					},
					"to": [
						{
							"shell_command": "pbpaste | pbcopy -pboard ruler"
						},
						{
							"key_code": "x",
							"modifiers": [
								"left_command"
							]
						},
						{
							"key_code": "open_bracket"
						},
						{
							"key_code": "v",
							"modifiers": [
								"left_command"
							]
						},
						{
							"shell_command": "pbpaste -pboard ruler | pbcopy"
						},
						{
							"key_code": "close_bracket"
						}
					],
					"type": "basic"
				}
			]
		}
	]
}
```
Problem remains: run this and the selected text still winds up in Jumpcut's clipboard history.

**Comment by Muirium** [comment]
Oh boy…

Experimenting with Control + K and Control + Y, I discover their behaviour is *app specific*. I was using BBEdit for my tests, and BBEdit syncs their clipboard with the general system pasteboard. TextEdit however keeps them apart. Safari considers them separate too. Scrivener is the same way. Oh, and all these apps (besides BBEdit) have separate pasteboards! So, soon enough I found myself with several simultaneous separate hidden pasteboards! Right now, there's different text pasting on Control Y in all 4!

So yeah, it's complicated. But better than nothing. I'll edit my macros to all use Control + K and Y. Most of my writing is not in BBEdit so should stay out of Jumpcut this way. Hopefully. Let's see!

**Comment by MuhammedZakir** [comment]
> Playing with pbpaste and pbcopy in Terminal just now, I see the ruler pasteboard is independent of Jumpcut's monitoring. I can throw stuff in there (an ls listing) and Jumpcut doesn't see it. So this should be possible. Can't say I understand what's going wrong.

The problem is with getting the currently selected text. I don't know any method for it except using clipboard.

**Comment by Muirium** [comment]
Yeah, me neither. Thanks for all your help though, it's been very useful indeed! Especially the one about Control Y and K. I vaguely remembered there was "another copy/paste on the Mac somewhere" but I had no idea how to go find it.

Control Y and K seems to be working nicely, outside of BBEdit. Maybe that's just a BBEdit bug. I'll know when Bare Bones get back to me.

I also just added `"repeat": false` to the last lines of my macros' `"to":` statements. Holding down keys was producing strange behaviour that's fixed now by suppressing repeat. Coincidentally, looking that function up, I rediscovered [Karabiner's documentation page](https://karabiner-elements.pqrs.org/docs/json/complex-modifications-manipulator-definition/to/repeat/) which clued me into macros existing in the first place! I just took that idea and ran with it here.

**Comment by MuhammedZakir** [comment]
To not make this a little easier, you could ask developer of your clipboard manager to add a cli or menu option to enable/disable/ignore clipboard watching. Then, you will be able to toggle it using shell script, or AppleScript.

**Comment by Muirium** [comment]
Thanks for reminding me to close this, bot!

I've been doing plenty with Karabiner macros, including single keystroke text selection and clipboard manipulation. For those interested, I've written loads about it here:

https://deskthority.net/viewtopic.php?p=494473#p494473

Macros are awesome!

**Comment by FloppyDisco** [comment]
VSCode has this functionality. and it does not mess with the clipboard history. i have no idea how they achieved this. it also works with all wrapping type characters: ( , { , [ , " , '
